web-api/src/personalization/personalized_knn.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2023 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::{collections::HashMap, time::Duration};
+
+use chrono::{DateTime, Utc};
+use futures_util::{stream::FuturesUnordered, Stream, StreamExt};
+use itertools::Itertools;
+use tracing::error;
+use xayn_ai_coi::{compute_coi_weights, nan_safe_f32_cmp, PositiveCoi};
+
+use crate::{
+    error::common::InternalError,
+    models::{DocumentId, PersonalizedDocument},
+    storage::{self, KnnSearchParams},
+    Error,
+};
+
+pub(super) struct Search<'a> {
+    pub(super) interests: &'a [PositiveCoi],
+    pub(super) excluded: &'a [DocumentId],
+    pub(super) horizon: Duration,
+    pub(super) max_cois: usize,
+    pub(super) count: usize,
+    pub(super) published_after: Option<DateTime<Utc>>,
+    pub(super) time: DateTime<Utc>,
+}
+
+impl Search<'_> {
+    /// Performs an approximate knn search for documents similar to the positive user interests.
+    pub(super) async fn run_on(
+        self,
+        storage: &impl storage::Document,
+    ) -> Result<Vec<PersonalizedDocument>, Error> {
+        let Search {
+            horizon,
+            max_cois,
+            count,
+            published_after,
+            interests,
+            excluded,
+            time,
+        } = self;
```

Comment:
because then I didn't need to "touch" the other code and have less changes, well except because it was also moved into it's own file and the indention level changed it's not making a difference :see_no_evil: